tapchannel/aux_invoice_manager_test.go

```diff
@@ -0,0 +1,798 @@
+// genRandomVertex generates a route.Vertex instance filled with random bytes.
+func genRandomVertex(t *rapid.T) route.Vertex {
+	var vertex route.Vertex
+	for i := 0; i < len(vertex); i++ {
+		vertex[i] = rapid.Byte().Draw(t, "vertex_byte")
+	}
+
+	return vertex
+}
+
+// genBuyQuotes populates the provided map of rfq quotes with the desired values
+// for a specific
+func genBuyQuotes(t *rapid.T, rfqMap rfq.BuyAcceptMap, units, amtMsat uint64,
+	scid rfqmsg.ID) {
+
+	// If the passed asset units is set to 0 this means that no wire custom
+	// records were set. To avoid a division by zero in the lines below we
+	// just set it to 1, this data is not relevant anymore.
+	if units == 0 {
+		units = 1
+	}
+
+	var peer route.Vertex
+	var assetRate *big.Int
+
+	// Introduce a chance that the quote's peerID is not correct.
+	if rapid.Bool().Draw(t, "nodeID_mismatch") {
+		peer = genRandomVertex(t)
+	} else {
+		peer = testNodeID
+	}
+
+	rfqScid := scid
+
+	// Introduce a chance that the quote's peerID is not correct.
+	if rapid.Bool().Draw(t, "scid_mismatch") {
+		rfqScid = rfqmsg.ID{2, 3, 4}
+	}
+
+	// Introduce a chance that the askPrice of this asset will result in a
+	// random total asset value.
+	if rapid.Bool().Draw(t, "no_asset_value_match") {
+		msatPerBtc := int64(btcutil.SatoshiPerBitcoin * 1000)
+		msatPerUnit := (float64(amtMsat) / float64(units)) + 1
+		assetRateInt := int64(float64(msatPerBtc)/msatPerUnit) + 1
+
+		// For a random asset unit price, draw a random value between
+		// 1/50th and double the expected price.
+		assetRate = big.NewInt(rapid.Int64Range(
+			assetRateInt/2, assetRateInt*50,
+		).Draw(t, "asset_msat_value"))
+	} else {
+		msatPerBtc := int64(btcutil.SatoshiPerBitcoin * 1000)
+		msatPerUnit := float64(amtMsat) / float64(units)
+		assetRate = big.NewInt(
+			int64(float64(msatPerBtc) / msatPerUnit),
+		)
+	}
+
+	rfqMap[rfqScid.Scid()] = rfqmsg.BuyAccept{
+		Peer: peer,
+		AssetRate: rfqmath.FixedPoint[rfqmath.BigInt]{
+			Coefficient: rfqmath.NewBigInt(assetRate),
+			Scale:       0,
+		},
+	}
+}
+
+// testInvoiceManager creates an array of requests to be processed by the
+// AuxInvoiceManager. Uses the enhanced HtlcModifierMockProperty instance.
+func testInvoiceManager(t *rapid.T) {
+	requests, rfqMap := genRequests(t)
+
+	mockRfq := &mockRfqManager{
+		peerBuyQuotes: rfqMap,
+	}
+
+	done := make(chan bool)
+
+	mockModifier := &mockHtlcModifierProperty{
+		requestQue: requests,
+		rfqMap:     rfqMap,
+		done:       done,
+		t:          t,
+	}
+
+	manager := NewAuxInvoiceManager(
+		&InvoiceManagerConfig{
+			ChainParams:         testChainParams,
+			InvoiceHtlcModifier: mockModifier,
+			RfqManager:          mockRfq,
+		},
+	)
+
+	err := manager.Start()
+	require.NoError(t, err)
+
+	select {
+	case <-done:
+	case <-time.After(testTimeout):
+		t.Fail()
+	}
+}
+
+// TestAuxInvoiceManagerProperty runs property based tests on the
+// AuxInvoiceManager.
+func TestAuxInvoiceManagerProperty(t *testing.T) {
+	t.Parallel()
```

Comment:
I think `rapid` tests are not thread safe? Not sure if that meant between multiple `rapid` tests though, or just any other test.

Reply:
Yeah, that's possible. But the `t.Parallel()` only means other tests in this file can run in parallel to the rapid test. And since we don't have any other rapid tests running there, this should be fine.